src/librustc_passes/ast_validation.rs

```diff
@@ -336,19 +336,24 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                     if let TraitItemKind::Method(ref sig, ref block) = trait_item.node {
                         self.check_trait_fn_not_async(trait_item.span, sig.header.asyncness);
                         self.check_trait_fn_not_const(sig.header.constness);
-                        if block.is_none() {
-                            self.check_decl_no_pat(&sig.decl, |span, mut_ident| {
-                                if mut_ident {
+                        self.check_decl_no_pat(&sig.decl, |span, mut_ident| {
+                            if mut_ident {
+                                if block.is_none() {
                                     self.session.buffer_lint(
                                         lint::builtin::PATTERNS_IN_FNS_WITHOUT_BODY,
                                         trait_item.id, span,
-                                        "patterns aren't allowed in methods without bodies");
-                                } else {
-                                    struct_span_err!(self.session, span, E0642,
-                                        "patterns aren't allowed in methods without bodies").emit();
+                                        "patterns aren't allowed in trait methods");
                                 }
-                            });
-                        }
+                            } else {
+                                let mut err = struct_span_err!(self.session, span, E0642,
```

Comment:
Ok, now this is a breaking change :smile: 
Some limited set of pattern (defined by `fn is_named_argument`) *was* previously accepted by parser in methods with body.
Something like
```rust
trait Tr {
    fn f(&ident: Type) {} // OK
    fn g(&&ident: Type) {} // OK
    fn h(mut ident: Type) {} // OK
    fn k($pat: Type) {} // OK
}
```
Note, that this compatibility check needs to happen in parser due to `$pat` that can look like an arbitrary complex pattern in AST validation, but still needs to be accepted.

Reply:
Oh no! I thought I accounted for that with the branches, but it turns out not! (Why did it even pass? Do we not have any tests for that?)

So will I need to change `parser.rs` to fix this, or can I just move the check for the body?

Reply:
I don't think AST validation needs any changes (it already does things right), all the relevant stuff is in the parser.
It would probably be better to:
- Check for `self.is_named_argument()` and parse in the old way if it's true.
- Otherwise, parse `PAT: TYPE` with state saving and immediately report a non-fatal error in case of success.
- Otherwise, backtrack and parse `TYPE`.

Reply:
If AST validation reports an unwanted duplicate error, then patterns could be converted into `_` in the parser.